torch/csrc/jit/passes/bailout_graph.cpp

```diff
@@ -0,0 +1,203 @@
+      buildBailOutBlockFrom(outer_node);
+    }
+  }
+
+  std::shared_ptr<Graph> buildBailOutGraphFrom(Node* n) {
+    old_to_new_.clear();
+    copy_graph_ = std::make_shared<Graph>();
+    buildBailOutBlockFrom(n);
+    // add graph outputs
+    for (auto ov : graph_->outputs()) {
+      auto new_value =
+          (this->old_to_new_.count(ov) == 0) ? nullptr : this->old_to_new_[ov];
+      auto nv = addNewInputForValue(ov, new_value);
+      copy_graph_->registerOutput(nv);
+    }
+    return copy_graph_;
+  }
+
+  std::shared_ptr<Graph> graph_;
+  std::shared_ptr<Graph> copy_graph_;
+  std::vector<Value*> live_inputs_;
+  std::unordered_map<Value*, Value*> old_to_new_;
+};
+
+struct BailOutInserter {
+  BailOutInserter(std::shared_ptr<Graph> graph) : graph_(std::move(graph)) {}
+
+  void run() {
+    insertBailOuts(graph_->block());
+    // set bailout subgraphs
+    setSubgraphs();
+  }
+
+  // remove all BailOuts and Guards
+  // we will profile this graph
+  // and insert guards and bailouts again
+  // where needed if its guard fails
+  void sanitizeGraph(Block* b) {
+    for (auto it = b->nodes().begin(); it != b->nodes().end(); ++it) {
+      if (it->kind() == prim::Guard) {
+        // this will need to be profiled again
+        it->input()->setType(TensorType::create());
+        // destroy the guard
+        it->output()->replaceAllUsesWith(it->input());
+        it.destroyCurrent();
+      }
+
+      if (it->kind() == prim::BailOut) {
+        // this is the index of an input we are actually guarding
+        auto index = it->i(attr::slot);
+        // this will need to be profiled again
+        it->inputs()[index]->setType(TensorType::create());
+        // destroy the guard
+        it->output()->replaceAllUsesWith(it->inputs()[index]);
+        it.destroyCurrent();
+      }
+
+      for (auto ib : it->blocks()) {
+        sanitizeGraph(ib);
+      }
+    }
+  }
+
+  void setSubgraphs() {
+    for (auto e : subgraphs) {
+      sanitizeGraph(e.second->block());
+      ConstantPooling(e.second);
+      e.first->g_(attr::Subgraph, e.second);
+    }
+  }
+
+  void insertBailOuts(Block* b) {
+    for (auto it = b->nodes().begin(); it != b->nodes().end(); ++it) {
+      if (it->kind() == prim::Guard) {
+        auto bailout_node = b->owningGraph()->create(prim::BailOut);
+        auto node = *it;
+
+        BailOutGraphBuilderForNode bg(graph_);
+        auto bailout_graph = bg.buildBailOutGraphFrom(node);
+
+        // N.B. we don't immediately set SubGraph attribute
+        // since we might need to copy this node the second time
+        // if the node is in a loop which then gets copied again and again
+        subgraphs.insert({bailout_node, bailout_graph});
+
+        for (size_t i = 0; i < bg.live_inputs_.size(); i++) {
+          bailout_node->addInput(bg.live_inputs_[i]);
+
+          // to tell which input (index offset)
+          // we are actually supposed to guard
+          if (it->input() == bg.live_inputs_[i]) {
```

Comment:
I don't like this organization because of how things will look in the future. BailOuts will be guarding multiple inputs and produce multiple outputs. The first N inputs/outputs should simply be the N things guarded.  Trying to allow it to be any input is going to be confusing to read and debug.

Reply:
In the future, when guards are coalesced together, it will guard many inputs, so it would be nice organize the node to anticipate that. Regardless of what happens here the actual interpreter support is going to need to handle the bailout inputs and the guarded inputs as separate things.

Reply:
I need to understand this use case a little bit better. I'll bug you about this tomorrow. One guarded input/one output is a very nice property to give up, tbh 😄